pkg/apis/pipeline/v1/matrix_types.go

```diff
@@ -96,11 +97,31 @@ func createCombination(name string, value string, combination Params) Params {
 
 // CountCombinations returns the count of combinations of Parameters generated from the Matrix in PipelineTask.
 func (m *Matrix) CountCombinations() int {
-	count := 0
-	if m.hasParams() {
-		count = 1
-		for _, param := range m.Params {
-			count *= len(param.Value.ArrayVal)
+	// Explicit combinations in the Matrix
+	// Calculate based on length of include Names and return count
+	if m.hasInclude() && !m.hasParams() {
+		return len(m.Include)
+	}
+	paramNames := make(map[string][]string)
+	count := 1
+	// Iterate over matrix.params and compute count of all generated combinations
+	for _, param := range m.Params {
+		count *= len(param.Value.ArrayVal)
+		paramNames[param.Name] = param.Value.ArrayVal
+	}
+	if m.hasInclude() {
+		// Check which combinations will be replaced vs generated based on matrix include params
+		for _, include := range m.Include {
+			for _, param := range include.Params {
+				if val, exist := paramNames[param.Name]; exist {
+					// If the matrix include param values does not exist, a new combination will be generated
+					if !slices.Contains(val, param.Value.StringVal) {
+						count++
+					} else {
+						break
+					}
+				}
+			}
```

Comment:
This is mainly to handle this kind of combination:

```
        - name: non-existent-arch
           params:
              - name: GOARCH
                value: "I-do-not-exist"
```

Reply:
but, let's say I defined the same `non-existent-arch` combination twice, would this count duplicates as two separate entities or counted as a single combination?

Reply:
@pritidesai  I don't think this would be valid to have the same include parameter defined twice..
@jerop correct me if I'm wrong.

Reply:
do we have such validation before reaching this point in the execution? If not, should we validate it or just de-dupe with a note to a user saying we have de-duped the combinations? I think we should be consistent with how params functions in case the same `param`  is defined twice. wdyt?

Reply:
I have created a PR implementing validation on duplicate matrix params https://github.com/tektoncd/pipeline/pull/6308

Reply:
Thanks Priti!

Reply:
Thanks for adding the validation @pritidesai!